File: back-end/src/main/java/es/ull/project/domain/entity/Execution.java
package es.ull.project.domain.entity;

import es.ull.project.domain.enumerate.ExecuteStatus;
import es.ull.project.domain.valueobject.File;
import es.ull.project.domain.valueobject.NodeExecuteData;
import es.ull.project.domain.valueobject.WorkFlowName;

import java.util.ArrayList;
import java.util.Collections;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

/**
 * @brief Value object representing the result of an execution.
 */
public class Execution {

    private static final String ERROR_STATUS_NOT_DEFINED = "Execute status not defined";
    private static final String ERROR_MESSAGE_NOT_DEFINED = "Message not defined";
    private static final String ERROR_OUTPUT_FILE_NOT_DEFINED = "Output file not defined";
    private static final String ERROR_NODES_EXECUTED_NOT_DEFINED = "Executed nodes list not defined";
    private static final String ERROR_TOTAL_EXECUTION_TIME_NOT_DEFINED = "Total execution time not defined";
    private static final String ERROR_NUMBER_OF_OPERATION_EXECUTED_NOT_DEFINED = "Number of operations executed not defined";
    private static final String ERROR_CREATION_DATE_NOT_DEFINED = "Creation date not defined";
    private static final String ERROR_WORKFLOW_NAME_NOT_DEFINED = "Workflow name not defined";
    
    private final UUID id;
    private final WorkFlowName workFlowName;
    private final LocalDateTime creationDate;
    
    private final ExecuteStatus status;
    private final String message;
    private final List<File> outputFiles;
    private final List<NodeExecuteData> nodesExecuted;
    private final Duration totalExecutionTime;
    private final int numberOfOperationExecuted;

    public Execution(
            WorkFlowName workFlowName,
            ExecuteStatus status,
            String message,
            List<File> outputFiles,
            List<NodeExecuteData> nodesExecuted,
            Duration totalExecutionTime,
            int numberOfOperationExecuted) {
        validateWorkFlowName(workFlowName);        
        validateStatus(status);
        validateMessage(message);
        validateOutputFile(outputFiles);
        validateNodesExecuted(nodesExecuted);
        validateTotalExecutionTime(totalExecutionTime);

        this.workFlowName = workFlowName;
        this.status = status;
        this.message = message;
        this.outputFiles = outputFiles;
        this.nodesExecuted = nodesExecuted;
        this.totalExecutionTime = totalExecutionTime;
        this.numberOfOperationExecuted = numberOfOperationExecuted;

        this.id = UUID.randomUUID();
        this.creationDate = LocalDateTime.now();
    }

    public Execution(Execution other) {
        this.id = other.id;
        this.workFlowName = other.workFlowName;
        this.creationDate = other.creationDate;
        this.status = other.status;
        this.message = other.message;
        this.outputFiles = new ArrayList<>(other.outputFiles);
        this.nodesExecuted = new ArrayList<>(other.nodesExecuted);
        this.totalExecutionTime = other.totalExecutionTime;
        this.numberOfOperationExecuted = other.numberOfOperationExecuted;
    }

    public Execution(
            UUID id,
            WorkFlowName workFlowName,
            LocalDateTime creationDate,
            ExecuteStatus status,
            String message,
            List<File> outputFiles,
            List<NodeExecuteData> nodesExecuted,
            Duration totalExecutionTime,
            int numberOfOperationExecuted) {
        validateWorkFlowName(workFlowName);
        validateStatus(status);
        validateMessage(message);
        validateOutputFile(outputFiles);
        validateNodesExecuted(nodesExecuted);
        validateTotalExecutionTime(totalExecutionTime);
        validateNumberOfOperationExecuted(numberOfOperationExecuted);
        validateCreationDate(creationDate);

        this.id = id;
        this.workFlowName = workFlowName;
        this.creationDate = creationDate;
        this.status = status;
        this.message = message;
        this.outputFiles = outputFiles;
        this.nodesExecuted = nodesExecuted;
        this.totalExecutionTime = totalExecutionTime;
        this.numberOfOperationExecuted = numberOfOperationExecuted;
    }

    public UUID getId() {
        return id;
    }

    public WorkFlowName getWorkFlowName() {
        return workFlowName;
    }

    public LocalDateTime getCreationDate() {
        return creationDate;
    }

    public ExecuteStatus getStatus() {
        return status;
    }

    public String getMessage() {
        return message;
    }

    public List<File> getOutputFiles() {
        return outputFiles;
    }

    public List<NodeExecuteData> getNodesExecuted() {
        return Collections.unmodifiableList(this.nodesExecuted);
    }

    public Duration getTotalExecutionTime() {
        return totalExecutionTime;
    }

    public int getNumberOfOperationExecuted() {
        return numberOfOperationExecuted;
    }

    public Execution setWorkFlowName(WorkFlowName workFlowName) {
        validateWorkFlowName(workFlowName);
        return new Execution(
                workFlowName,
                this.status,
                this.message,
                this.outputFiles,
                this.nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setStatus(ExecuteStatus status) {
        validateStatus(status);
        return new Execution(
                this.workFlowName,
                status,
                this.message,
                this.outputFiles,
                this.nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setMessage(String message) {
        validateMessage(message);
        return new Execution(
                this.workFlowName,
                this.status,
                message,
                this.outputFiles,
                this.nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setOutputFile(File outputFile) {
        validateOutputFile(List.of(outputFile));
        return new Execution(
                this.workFlowName,
                this.status,
                this.message,
                List.of(outputFile),
                this.nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setOutputFiles(List<File> outputFiles) {
        validateOutputFile(outputFiles);
        return new Execution(
                this.workFlowName,
                this.status,
                this.message,
                outputFiles,
                this.nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setNodesExecuted(List<NodeExecuteData> nodesExecuted) {
        validateNodesExecuted(nodesExecuted);
        return new Execution(
                this.workFlowName,
                this.status,
                this.message,
                this.outputFiles,
                nodesExecuted,
                this.totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setTotalExecutionTime(Duration totalExecutionTime) {
        validateTotalExecutionTime(totalExecutionTime);
        return new Execution(
                this.workFlowName,
                this.status,
                this.message,
                this.outputFiles,
                this.nodesExecuted,
                totalExecutionTime,
                this.numberOfOperationExecuted);
    }

    public Execution setNumberOfOperationExecuted(int numberOfOperationExecuted) {
        validateNumberOfOperationExecuted(numberOfOperationExecuted);
        return new Execution(
                this.workFlowName,
                this.status,
                this.message,
                this.outputFiles,
                this.nodesExecuted,
                this.totalExecutionTime,
                numberOfOperationExecuted);
    }


    private void validateWorkFlowName(WorkFlowName workFlowName) {
        if (workFlowName == null) {
            throw new IllegalArgumentException(ERROR_WORKFLOW_NAME_NOT_DEFINED);
        }
    }

    private void validateCreationDate(LocalDateTime creationDate) {
        if (creationDate == null) {
            throw new IllegalArgumentException(ERROR_CREATION_DATE_NOT_DEFINED);
        }
    }
    private void validateStatus(ExecuteStatus status) {
        if (status == null) {
            throw new IllegalArgumentException(ERROR_STATUS_NOT_DEFINED);
        }
    }

    private void validateMessage(String message) {
        if (message == null || message.isBlank()) {
            throw new IllegalArgumentException(ERROR_MESSAGE_NOT_DEFINED);
        }
    }

    private void validateOutputFile(List<File> outputFile) {
        if (outputFile == null) {
            throw new IllegalArgumentException(ERROR_OUTPUT_FILE_NOT_DEFINED);
        }
    }

    private void validateNodesExecuted(List<NodeExecuteData> nodesExecuted) {
        if (nodesExecuted == null) {
            throw new IllegalArgumentException(ERROR_NODES_EXECUTED_NOT_DEFINED);
        }
    }

    private void validateTotalExecutionTime(Duration totalExecutionTime) {
        if (totalExecutionTime == null) {
            throw new IllegalArgumentException(ERROR_TOTAL_EXECUTION_TIME_NOT_DEFINED);
        }
    }

    private void validateNumberOfOperationExecuted(int numberOfOperationExecuted) {
        if (numberOfOperationExecuted < 0) {
            throw new IllegalArgumentException(ERROR_NUMBER_OF_OPERATION_EXECUTED_NOT_DEFINED);
        }
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof Execution)) {
            return false;
        }
        Execution that = (Execution) otherObject;
        return (status.equals(that.status)) &&
                Objects.equals(message, that.message) &&
                Objects.equals(outputFiles, that.outputFiles) &&
                Objects.equals(nodesExecuted, that.nodesExecuted) &&
                Objects.equals(totalExecutionTime, that.totalExecutionTime) &&
                (numberOfOperationExecuted == that.numberOfOperationExecuted);
    }

    @Override
    public int hashCode() {
        return Objects.hash(status, message, outputFiles, nodesExecuted, totalExecutionTime, numberOfOperationExecuted);
    }

    @Override
    public String toString() {
        return String.format(
                "ExecuteData={id=%s, workflowName=%s, creationDate=%s status=%s, message='%s', outputFiles=%s, nodesExecuted=%s, totalExecutionTime=%s, numberOfOperationExecuted=%d}",
                id, workFlowName, creationDate, status, message, outputFiles, nodesExecuted, totalExecutionTime, numberOfOperationExecuted);
    }
}

File: front-end/src/domain/entity/execution.ts
import { UllUUID } from "@ull-tfg/ull-tfg-typescript";
import { ExecuteStatus } from "../enumerate/ExecuteStatus";
import { File } from "../value-object/file";
import { NodeExecutedData } from "../value-object/node-execute-data";
import { WorkFlowName } from "../value-object/workflow-name";

export class Execution {
    private static readonly ERROR_STATUS_NOT_DEFINED = "Execute status not defined";
    private static readonly ERROR_MESSAGE_NOT_DEFINED = "Message not defined";
    private static readonly ERROR_OUTPUT_FILE_NOT_DEFINED = "Output files not defined";
    private static readonly ERROR_NODES_EXECUTED_NOT_DEFINED = "Executed nodes list not defined";
    private static readonly ERROR_TOTAL_EXECUTION_TIME_NOT_DEFINED = "Total execution time must be a positive number";
    private static readonly ERROR_ID_NOT_DEFINED = "ID not defined";
    private static readonly ERROR_ID_INVALID = "ID must be an instance of UllUUID";

    private static readonly ERROR_WORKFLOW_NAME_NOT_DEFINED = "Workflow name not defined";

    private readonly id: UllUUID;
    private readonly workFlowName: WorkFlowName;
    private readonly creationDate: Date;

    private readonly status: ExecuteStatus;
    private readonly message: string;
    private readonly outputFiles: File[];
    private readonly nodesExecuted: NodeExecutedData[];
    private readonly totalExecutionTimeMillis: number;
    private readonly numberOfOperationExecuted: number;

    constructor(
        workFlowName: WorkFlowName,
        status: ExecuteStatus,
        message: string,
        outputFiles: File[],
        nodesExecuted: NodeExecutedData[],
        totalExecutionTimeMillis: number,
        numberOfOperationExecuted: number,
        creationDate: Date | string = new Date(), id?: UllUUID
    ) {
        this.validateWorkFlowName(workFlowName);
        this.validateStatus(status);
        this.validateMessage(message);
        this.validateOutputFiles(outputFiles);
        this.validateNodesExecuted(nodesExecuted);
        this.validateTotalExecutionTime(totalExecutionTimeMillis);

        if (id) {
            this.validateId(id);
            this.id = id;
        } else {
            this.id = UllUUID.random();
        }

        this.workFlowName = workFlowName;
        this.creationDate = new Date(creationDate);
        this.status = status;
        this.message = message;
        this.outputFiles = outputFiles;
        this.nodesExecuted = nodesExecuted;
        this.totalExecutionTimeMillis = totalExecutionTimeMillis;
        this.numberOfOperationExecuted = numberOfOperationExecuted;
    }

    getId(): UllUUID {
        return this.id;
    }

    getWorkFlowName(): WorkFlowName {
        return this.workFlowName;
    }

    getCreationDate(): Date {
        return this.creationDate;
    }

    getStatus(): ExecuteStatus {
        return this.status;
    }

    getMessage(): string {
        return this.message;
    }

    getOutputFiles(): File[] {
        return this.outputFiles;
    }

    getNodesExecuted(): NodeExecutedData[] {
        return this.nodesExecuted;
    }

    getTotalExecutionTimeMillis(): number {
        return this.totalExecutionTimeMillis;
    }

    getNumberOfOperationExecuted(): number {
        return this.numberOfOperationExecuted;
    }

    setWorkFlowName(workFlowName: WorkFlowName): Execution {
        this.validateWorkFlowName(workFlowName);
        return new Execution(workFlowName, this.status, this.message, this.outputFiles, this.nodesExecuted, this.totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setStatus(status: ExecuteStatus): Execution {
        return new Execution(this.workFlowName, status, this.message, this.outputFiles, this.nodesExecuted, this.totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setMessage(message: string): Execution {
        return new Execution(this.workFlowName, this.status, message, this.outputFiles, this.nodesExecuted, this.totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setOutputFiles(outputFiles: File[]): Execution {
        return new Execution(this.workFlowName, this.status, this.message, outputFiles, this.nodesExecuted, this.totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setNodesExecuted(nodesExecuted: NodeExecutedData[]): Execution {
        return new Execution(this.workFlowName, this.status, this.message, this.outputFiles, nodesExecuted, this.totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setTotalExecutionTimeMillis(totalExecutionTimeMillis: number): Execution {
        return new Execution(this.workFlowName, this.status, this.message, this.outputFiles, this.nodesExecuted, totalExecutionTimeMillis, this.numberOfOperationExecuted);
    }

    setNumberOfOperationExecuted(numberOfOperationExecuted: number): Execution {
        return new Execution(this.workFlowName, this.status, this.message, this.outputFiles, this.nodesExecuted, this.totalExecutionTimeMillis, numberOfOperationExecuted);
    }

    private validateId(id: UllUUID): void {
        if (!id) throw new Error(Execution.ERROR_ID_NOT_DEFINED);
        if (!(id instanceof UllUUID)) throw new Error(Execution.ERROR_ID_INVALID);
    }

    private validateWorkFlowName(workFlowName: WorkFlowName): void {
        if (!workFlowName || workFlowName.toString().trim() === "") {
            throw new Error(Execution.ERROR_WORKFLOW_NAME_NOT_DEFINED);
        }
    }
    private validateStatus(status: ExecuteStatus): void {
        if (status === null || status === undefined) {
            throw new Error(Execution.ERROR_STATUS_NOT_DEFINED);
        }
    }

    private validateMessage(message: string): void {
        if (!message || message.trim() === "") {
            throw new Error(Execution.ERROR_MESSAGE_NOT_DEFINED);
        }
    }

    private validateOutputFiles(outputFile: File[]): void {
        if (!outputFile) {
            throw new Error(Execution.ERROR_OUTPUT_FILE_NOT_DEFINED);
        }
    }

    private validateNodesExecuted(nodesExecuted: NodeExecutedData[]): void {
        if (!nodesExecuted) {
            throw new Error(Execution.ERROR_NODES_EXECUTED_NOT_DEFINED);
        }
    }

    private validateTotalExecutionTime(totalExecutionTimeMillis: number): void {
        if (totalExecutionTimeMillis == null || totalExecutionTimeMillis < 0) {
            throw new Error(Execution.ERROR_TOTAL_EXECUTION_TIME_NOT_DEFINED);
        }
    }

    toString(): string {
        return `Execution{id='${this.id}', workFlowName='${this.workFlowName}', creationDate='${this.creationDate}', status='${this.status}', message='${this.message}', outputFiles=${this.outputFiles.toString()}, nodesExecuted=[${this.nodesExecuted.map(n => n.toString()).join(", ")}], totalExecutionTimeMillis=${this.totalExecutionTimeMillis}, numberOfOperationExecuted=${this.numberOfOperationExecuted}}`;
    }

    equals(other: unknown): boolean {
        if (!(other instanceof Execution)) return false;
        return (
            this.workFlowName.equals(other.workFlowName) &&
            this.id.equals(other.id) &&
            this.creationDate.getTime() === other.creationDate.getTime() &&
            this.status === other.status &&
            this.message === other.message &&
            this.arraysEqualFiles(this.outputFiles, other.outputFiles) &&
            this.arraysEqualNodeName(this.nodesExecuted, other.nodesExecuted) &&
            this.totalExecutionTimeMillis === other.totalExecutionTimeMillis
        );
    }

    private arraysEqualNodeName(arr1: NodeExecutedData[], arr2: NodeExecutedData[]): boolean {
        if (arr1.length !== arr2.length) return false;
        for (let i = 0; i < arr1.length; i++) {
            if (!arr1[i].equals(arr2[i])) return false;
        }
        return true;
    }

    private arraysEqualFiles(arr1: File[], arr2: File[]): boolean {
        if (arr1.length !== arr2.length) return false;
        for (let i = 0; i < arr1.length; i++) {
            if (!arr1[i].equals(arr2[i])) return false;
        }
        return true;
    }

    // Simple hash code based on JSON string (not perfect but illustrative)
    hashCode(): number {
        let str = JSON.stringify({
            status: this.status,
            message: this.message,
            outputFile: this.outputFiles.toString(),
            nodesExecuted: this.nodesExecuted.map(n => n.toString()),
            totalExecutionTimeMillis: this.totalExecutionTimeMillis,
        });
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            const chr = str.charCodeAt(i);
            hash = (hash << 5) - hash + chr;
            hash |= 0;
        }
        return hash;
    }
}
